File: sdt-perl/step0_repl.pl
#!/usr/bin/env perl

use 5.20.1;
use warnings;

use Function::Parameters qw( :strict );
use Term::ReadLine;

exit main(@ARGV) ? 0 : 1;

#-------------------------------------------------------------------------------

fun main(@argv) {
    my $term = Term::ReadLine->new('mal');
    my $prompt = 'user>';
    while (defined(my $input = $term->readline($prompt))) {
        chomp($input);
        say rep($input);
        $term->addhistory($input) if $input =~ /\S/;
    }
    return 0;
}

fun rep($arg) {
    return PRINT(EVAL(READ($arg)));
}

fun READ($arg) {
    return $arg;
}

fun EVAL($arg) {
    return $arg;
}

fun PRINT($arg) {
    return $arg;
}

File: sdt-cpp/step0_repl.cpp
#include "MAL.h"

#include <iostream>

int main(int argc, char* argv[])
{
    String prompt = "user> ";
    while (1) {
        String input;
        std::cout << prompt;
        std::getline(std::cin, input);
        if (std::cin.eof() || std::cin.fail()) {
            break;
        }

        std::cout << rep(input) << "\n";
    }
}

String rep(const String& input)
{
    return PRINT(EVAL(READ(input)));
}

String READ(const String& input)
{
    return input;
}

String EVAL(const String& input)
{
    return input;
}

String PRINT(const String& input)
{
    return input;
}
